src/token/src/canister.rs

```diff
@@ -207,38 +207,59 @@ impl TokenCanister {
     /// is not the global range of all transactions. The range [start, start + limit) here pertains to
     /// the transactions of user who. Implementations are allowed to return less TxRecords than
     /// requested to fend off DoS attacks.
+    ///
+    /// # Arguments
+    /// * `who` - The user to get transactions for.
+    /// * `start` - The index of the first transaction to return.
+    /// * `limit` - The number of transactions to return.
     #[query]
     fn getUserTransactions(&self, who: Principal, start: Nat, limit: Nat) -> Vec<TxRecord> {
-        let mut transactions = vec![];
-
         let limit_usize = limit.0.to_usize().unwrap_or(usize::MAX);
         if limit_usize > MAX_TRANSACTION_QUERY_LEN {
             ic_kit::ic::trap(&format!(
-                "Limit must be less then {}",
+                "Limit must be less than {}",
                 MAX_TRANSACTION_QUERY_LEN
             ));
         }
 
-        for tx in self.state.borrow().ledger.get_range(&start, &limit) {
-            if tx.from == who || tx.to == who || tx.caller == Some(who) {
-                transactions.push(tx.clone());
-            }
+        let length = self
+            .state
+            .borrow()
+            .ledger
+            .iter()
+            .filter(|tx| tx.from == who || tx.to == who || tx.caller == Some(who))
+            .count();
+
+        if start > length {
+            ic_kit::ic::trap(&format!("Start must be less than {}", length));
         }
 
+        let mut transactions: Vec<TxRecord> = self
+            .state
+            .borrow()
+            .ledger
+            .iter()
+            .filter(|tx| tx.from == who || tx.to == who || tx.caller == Some(who))
+            .skip(start.0.to_usize().unwrap())
+            .take(limit_usize)
+            .cloned()
+            .collect();
+
+        transactions.reverse();
+
         transactions
     }
-
-    /// Returns total number of transactions related to the user `who`.
+    /// Returns the total number of transactions related to the user `who`.
     #[query]
-    fn getUserTransactionAmount(&self, who: Principal) -> Nat {
-        let mut amount = Nat::from(0);
+    fn getUserTransactionCount(&self, who: Principal) -> Nat {
```

Comment:
This function is never going to do an overflowing arithmetic operation.
It can be rewritten to use a `usize` instead since `Nat` makes heap allocations and we can limit the number of heap allocations to just one:

```rust
    /// Returns the total number of transactions related to the user `who`.
    #[query]
    fn getUserTransactionCount(&self, who: Principal) -> Nat {
        let mut count = 0usize;
        for tx in self.state.borrow().ledger.iter() {
            if tx.from == who || tx.to == who || tx.caller == Some(who) {
                count += 1;
            }
        }

        Nat::from(count)
    }
```